### ISSUE TITLE
Allow connections to an httptest Server

### ISSUE THREAD

**Issue by michele** [h2non/gock#18]
Hi,

I'm finding gock extremely useful, however I have a test where I need to make a real call to an httptest Server which will then spawn a couple of requests which gock should intercept.
However, I can't seem to find a way to do it. I tried enabling networking and filtering, which makes it work, but it won't call the actual server, it just mocks that call, too.

```golang
It("receives a complete trigger [DOC]", func() {
  defer gock.DisableNetworking()
  defer gock.DisableNetworkingFilters()
  gock.EnableNetworking()
  var a args
  a.Host = docServer.URL
  a.Path = "/triggers"
  a.Body = completeTrigger
  a.Method = "POST"
  a.Headers = map[string]string{
    "Authorization": "Token token=" + configuration.Config.AppToken,
    "Content-Type":  middleware.MimeType,
    "Accept":        middleware.MimeType,
  }
  gock.NetworkingFilter(func(req *http.Request) bool {
    return req.URL.String() == docServer.URL
  })
  gock.New(docServer.URL).
    Post("/triggers").
    Reply(201)

  // here are the actual mocks I need

  resp, err := a.run()
  Expect(err).ToNot(HaveOccurred())
  Expect(resp.StatusCode).To(Equal(201))
  Expect(gock.IsDone())
})
```

The first Expect will generate this error: 

```
Expected error:
    <*url.Error | 0xc4204040f0>: {
        Op: "Post",
        URL: "http://127.0.0.1:59186/triggers",
        Err: {
            s: "gock: cannot match any request",
        },
    }
    Post http://127.0.0.1:59186/triggers: gock: cannot match any request
not to have occurred
```

Am I missing something?

Thanks a lot for your help!

**Comment by michele** [comment]
I made another attempt:

```golang
gock.New(docServer.URL).
    EnableNetworking()
```

And I get this:

```
    Expected error:
        <*url.Error | 0xc4203b99b0>: {
            Op: "Post",
            URL: "http://127.0.0.1:62537/triggers",
            Err: {
                err: "net/http: request canceled (Client.Timeout exceeded while awaiting headers)",
                timeout: true,
            },
        }
        Post http://127.0.0.1:62537/triggers: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
    not to have occurred
```

**Comment by michele** [comment]
I got it! What was happening is that the httptest Server was making calls that needed to be mocked, but since the mutex was locked, gock wasn't able to serve those and the Server would be waiting until timeout. I fixed it by unlocking the mutex if networking is enabled and mock is nil. Only inconvenience is that the unlock can no longer be deferred or a panic would be raised in case of a double unlock.

I can send you a pull request if you're interested.

**Comment by cintosyntax** [comment]
Hey @michele, I am trying to grasp the problem you are having. Do you have multiple tests that sometimes `DisableNetwork` and other `EnableNetwork`. Based on your solution, I think the state of the network setting on `gock` is being changed mid execution of your test example from other tests running in parallel.

Am I correct about this? Thanks =)

**Comment by maxcnunes** [comment]
@h2non You can close it now. This have been fixed with your changes to make https://github.com/h2non/gock/pull/23 work.

**Comment by h2non** [comment]
That's correct. Closing this.
